hoodie-client/src/main/java/com/uber/hoodie/io/HoodiePartitionHelper.java

```diff
@@ -0,0 +1,112 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.hoodie.io;
+
+import com.uber.hoodie.common.model.HoodieRecord;
+import com.uber.hoodie.table.WorkloadProfile;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+import scala.Tuple2;
+import scala.Tuple3;
+
+/**
+ * HoodiePartitionHelper is responsible for splitting Hoodie upsert stage into smaller spark
+ * stages.
+ */
+public class HoodiePartitionHelper {
+
+  /**
+   * Split an upsert workload
+   *
+   * @param profile WorkloadProfile for the input record
+   * @param isUpsert indicates upsert workflow or insert only workflow
+   * @param maxPartitions max spark partitions allowed for single datafile write stage
+   * @return List of conditions in Tuple3 of (lowBound, highBound, count)
+   */
+  public static List<Tuple3<String, String, Long>> splitWorkload(WorkloadProfile profile,
+      boolean isUpsert,
+      int maxPartitions) {
+
+    // We can't split input records if we don't know the workload or if this is a plain insert
+    if (profile == null || !isUpsert) {
+      return Collections.emptyList();
+    }
+
+    Set<String> sortedPartitions = new TreeSet<>();
+    List<Tuple2<String, Long>> splitPoints = new ArrayList<>();
+    sortedPartitions.addAll(profile.getPartitionPaths());
+    long totalPartitions = 0;
+    for (String partition : sortedPartitions) {
+      int updateCount = profile.getWorkloadStat(partition)
+          .getUpdateLocationToCount().size();
+      if (totalPartitions == 0
+          || totalPartitions + updateCount <= maxPartitions) {
+        totalPartitions += updateCount;
+      } else {
+        splitPoints.add(new Tuple2<>(partition, totalPartitions));
+        totalPartitions = updateCount;
+      }
+    }
+
+    if (splitPoints.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    // Figure filter conditions
+    List<Tuple3<String, String, Long>> conditions = new ArrayList<>();
+    conditions.add(new Tuple3<>(null, splitPoints.get(0)._1(), splitPoints.get(0)._2()));
+    for (int i = 0; i < splitPoints.size(); ++i) {
+      if (i == splitPoints.size() - 1) {
+        conditions.add(new Tuple3<>(splitPoints.get(i)._1(), null, totalPartitions));
+      } else {
+        conditions.add(new Tuple3<>(splitPoints.get(i)._1(), splitPoints.get(i + 1)._1(),
+            splitPoints.get(i + 1)._2()));
+      }
+    }
+    return conditions;
+  }
+
+  /**
+   * Filter input records by the given condition
+   *
+   * @param hoodieRecord HoodieRecord to filter
+   * @param condition the condition used to filter a given HoodieRecord
+   * @return a Boolean indicates the filter result
+   */
+  public static Boolean filterHoodieRecord(HoodieRecord hoodieRecord,
```

Comment:
This looks confusing to me, can way you can make this easier to understand ?

Reply:
I can introduce a Pojo for the tuple to capture lower bound and higher bound.  @vinothchandar does it sounds good to you?  I was leaning towards not to introducing an extra class.